Add NCycle iterator

lstagner opened the pull request:
I think it is useful to be able to set how many times an iterator should be cycled over.

``` julia
julia> for i=cycle(1:2,2)
           println(i)
       end
1
2
1
2
```

The documentation should be changed but I am unsure of where to change it.

Commit:
Created NCycle iterator type

tkelman commented:
The docstring for this is currently at https://github.com/JuliaLang/julia/blob/3f6b2b26a2d4ccef77f0b54ae664e5dcae157db1/base/docs/helpdb/Base.jl#L5565-L5570, but this would be a good opportunity to move it inline next to the source code. If you change the signature in the docstring, you should make the corresponding change to the signature in `doc/stdlib/collections.rst` then run `make docs` so the content of the docstring gets translated over.

lstagner commented:
I have added tests and documentation. I have also moved the `cycle` docs inline as you suggested.

JeffBezanson commented:
This doesn't work as well as it could since it requires `xs` to support `length`.

lstagner commented:
Do you have a suggestion on how to improve it? The only way I know of is to create a new iterator type.

kshyatt commented:
Well, well, this is your first Julia PR, isn't it @lstagner? Welcome to Julia!

If you could run the doctests as well `make -C doc doctest` to make sure they all still pass, that would be awesome.

lstagner commented:
@kshyatt Its actually my third. A few years ago I added tab completion for superscripts and subscripts https://github.com/JuliaLang/julia/pull/6927 https://github.com/JuliaLang/julia/pull/7018

I ran `make -C doc doctest` and there were three failures. All unrelated to this PR
There were two errors in `stdlib/linalg.rst` and one error in `stdlib/collections.rst` 
 
```
Document: stdlib/linalg
-----------------------
**********************************************************************
File "stdlib/linalg.rst", line 651, in default
Failed example:
    eigmax(A)
Expected:
    ERROR: DomainError:
     in #eigmax#30(::Bool, ::Bool, ::Function, ::Array{Complex{Int64},2}) at ./linalg/eigen.jl:108
     in eigmax(::Array{Complex{Int64},2}) at ./linalg/eigen.jl:106
Got:
    ERROR: DomainError:
     in #eigmax#30(::Bool, ::Bool, ::Function, ::Array{Complex{Int64},2}) at ./linalg/eigen.jl:185
     in eigmax(::Array{Complex{Int64},2}) at ./linalg/eigen.jl:183
     in eval(::Module, ::Any) at ./boot.jl:234
     in eval_user_input(::Any, ::Bool) at ./client.jl:117
     in eval(::Module, ::Any) at ./boot.jl:234
     in eval_user_input(::Any, ::Bool) at ./client.jl:117
     in _start() at ./client.jl:355
**********************************************************************
File "stdlib/linalg.rst", line 677, in default
Failed example:
    eigmin(A)
Expected:
    ERROR: DomainError:
     in #eigmin#31(::Bool, ::Bool, ::Function, ::Array{Complex{Int64},2}) at ./linalg/eigen.jl:115
     in eigmin(::Array{Complex{Int64},2}) at ./linalg/eigen.jl:113
Got:
    ERROR: DomainError:
     in #eigmin#31(::Bool, ::Bool, ::Function, ::Array{Complex{Int64},2}) at ./linalg/eigen.jl:224
     in eigmin(::Array{Complex{Int64},2}) at ./linalg/eigen.jl:222
     in eval(::Module, ::Any) at ./boot.jl:234
     in eval_user_input(::Any, ::Bool) at ./client.jl:117
     in eval(::Module, ::Any) at ./boot.jl:234
     in eval_user_input(::Any, ::Bool) at ./client.jl:117
     in _start() at ./client.jl:355
**********************************************************************
1 items had failures:
   2 of  17 in default
17 tests in 1 items.
15 passed and 2 failed.
***Test Failed*** 2 failures.
```

and
```
Document: stdlib/collections
----------------------------
**********************************************************************
File "stdlib/collections.rst", line 1379, in default
Failed example:
    deleteat!([6, 5, 4, 3, 2, 1], (2, 2))
Expected:
    ERROR: ArgumentError: indices must be unique and sorted
     in deleteat!(::Array{Int64,1}, ::Tuple{Int64,Int64}) at ./array.jl:576
     ...
Got:
    ERROR: ArgumentError: indices must be unique and sorted
     in deleteat!(::Array{Int64,1}, ::Tuple{Int64,Int64}) at ./array.jl:611
     in eval(::Module, ::Any) at ./boot.jl:234
     in eval_user_input(::Any, ::Bool) at ./client.jl:117
     in eval(::Module, ::Any) at ./boot.jl:234
     in eval_user_input(::Any, ::Bool) at ./client.jl:117
     in _start() at ./client.jl:355
**********************************************************************
1 items had failures:
   1 of  94 in default
94 tests in 1 items.
93 passed and 1 failed.
***Test Failed*** 1 failures.
```

lstagner commented:
I have done a rebase against master and all the unrelated doc test failures now pass. 

This is ready to merge whenever.

lstagner commented:
This PR wasn't getting much traction so I decided to create a custom iterator type instead of using `take`. 

```julia
julia> for i=ncycle(1:2,2)
           println(i)
       end
1
2
1
2
```

lstagner commented:
I have rebased and this is ready for review/merging.

I addressed @JeffBezanson comment by making a new iterator type `NCycle` instead of using `take`.

iamed2 commented:
As a new iterator this might fit better in [Iterators.jl](https://github.com/JuliaLang/Iterators.jl).

JeffBezanson commented:
Are there use cases for this that justify putting it in Base? The existing `cycle` iterator itself doesn't seem to be used in Base, and with #5155 these things are drawing closer scrutiny.

lstagner commented:
If you think Iterators.jl is a better fit for this I will open up a PR there.

lstagner commented:
https://github.com/JuliaLang/Iterators.jl/pull/88
